### ISSUE TITLE
PutRecord bytes sum very low

### ISSUE THREAD

**Issue by Brico87** [spring-cloud/spring-cloud-stream-binder-aws-kinesis#65]
We experienced some issues with the Spring Cloud Stream binder for AWS Kinesis at version **1.0.0.RELEASE**. We have implemented a simple test case with a source, a sink and a Kinesis stream between.

The data producer is the Spring Cloud Stream Time Source (https://github.com/spring-cloud/spring-cloud-stream-modules/tree/master/time-source).

The data consumer is the Throughput Sink (https://github.com/spring-cloud/spring-cloud-stream-modules/tree/master/throughput-sink).

We tested on Windows 10, Java 8 and Kinesis stream emulated by Localstack (Version 0.8.7). We also tested with an online stream and EC2 deployments and with mono-shard and multi-shards streams.

These are log extracts of the throughtput sink application for time source configured with 1000 messages max per second. The time-source jar and the sink jar are running on two separate EC2 machines :

```bash
$ java -Dcloud.aws.stack.auto=false -jar throughput-sink-1.0.0.BUILD-SNAPSHOT-exec.jar

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.0.4.RELEASE)

2018-10-09 14:11:21.868  INFO 3107 --- [           main] o.s.c.s.m.t.ThroughputSinkApplication    : Starting ThroughputSinkApplication v1.0.0.BUILD-SNAPSHOT on ip-XXX-XX-XXX with PID 3107 (/home/ec2-user/throughput-sink-1.0.0.BUILD-SNAPSHOT-exec.jar started by ec2-user in /home/ec2-user)
2018-10-09 14:11:21.885  INFO 3107 --- [           main] o.s.c.s.m.t.ThroughputSinkApplication    : No active profile set, falling back to default profiles: default
2018-10-09 14:11:22.038  INFO 3107 --- [           main] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@5c0369c4: startup date [Tue Oct 09 14:11:22 UTC 2018]; root of context hierarchy
2018-10-09 14:11:23.413  INFO 3107 --- [           main] o.s.i.config.IntegrationRegistrar        : No bean named 'integrationHeaderChannelRegistry' has been explicitly defined. Therefore, a default DefaultHeaderChannelRegistry will be created.
2018-10-09 14:11:23.997  INFO 3107 --- [           main] faultConfiguringBeanFactoryPostProcessor : No bean named 'errorChannel' has been explicitly defined. Therefore, a default PublishSubscribeChannel will be created.
2018-10-09 14:11:24.008  INFO 3107 --- [           main] faultConfiguringBeanFactoryPostProcessor : No bean named 'taskScheduler' has been explicitly defined. Therefore, a default ThreadPoolTaskScheduler will be created.
2018-10-09 14:11:24.224  INFO 3107 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#4d3167f4' of type [com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-10-09 14:11:24.245  INFO 3107 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean '(inner bean)#4eb7f003' of type [com.amazonaws.auth.profile.ProfileCredentialsProvider] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-10-09 14:11:24.266  INFO 3107 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'credentialsProvider' of type [org.springframework.cloud.aws.core.credentials.CredentialsProviderFactoryBean] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-10-09 14:11:24.266  INFO 3107 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'credentialsProvider' of type [com.amazonaws.auth.AWSCredentialsProviderChain] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-10-09 14:11:24.279  WARN 3107 --- [           main] com.amazonaws.util.EC2MetadataUtils      : Unable to retrieve the requested metadata.
2018-10-09 14:11:24.435  INFO 3107 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'org.springframework.integration.config.IntegrationManagementConfiguration' of type [org.springframework.integration.config.IntegrationManagementConfiguration$$EnhancerBySpringCGLIB$$ca33f8a3] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
2018-10-09 14:11:25.419  INFO 3107 --- [           main] o.s.s.c.ThreadPoolTaskScheduler          : Initializing ExecutorService  'taskScheduler'
2018-10-09 14:11:27.542  INFO 3107 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Registering beans for JMX exposure on startup
2018-10-09 14:11:27.559  INFO 3107 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Bean with name 'integrationMbeanExporter' has been autodetected for JMX exposure
2018-10-09 14:11:27.564  INFO 3107 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Located managed bean 'integrationMbeanExporter': registering with JMX server as MBean [org.springframework.integration.monitor:name=integrationMbeanExporter,type=IntegrationMBeanExporter]
2018-10-09 14:11:27.679  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering beans for JMX exposure on startup
2018-10-09 14:11:27.679  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageChannel errorChannel
2018-10-09 14:11:27.692  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Located managed bean 'org.springframework.integration:type=MessageChannel,name=errorChannel': registering with JMX server as MBean [org.springframework.integration:type=MessageChannel,name=errorChannel]
2018-10-09 14:11:27.811  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageChannel nullChannel
2018-10-09 14:11:27.814  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Located managed bean 'org.springframework.integration:type=MessageChannel,name=nullChannel': registering with JMX server as MBean [org.springframework.integration:type=MessageChannel,name=nullChannel]
2018-10-09 14:11:27.836  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageChannel input
2018-10-09 14:11:27.838  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Located managed bean 'org.springframework.integration:type=MessageChannel,name=input': registering with JMX server as MBean [org.springframework.integration:type=MessageChannel,name=input]
2018-10-09 14:11:27.861  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageHandler errorLogger
2018-10-09 14:11:27.862  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Located managed bean 'org.springframework.integration:type=MessageHandler,name=errorLogger,bean=internal': registering with JMX server as MBean [org.springframework.integration:type=MessageHandler,name=errorLogger,bean=internal]
2018-10-09 14:11:27.906  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Registering MessageHandler throughputSink.throughputSink.serviceActivator
2018-10-09 14:11:27.907  INFO 3107 --- [           main] o.s.i.monitor.IntegrationMBeanExporter   : Located managed bean 'org.springframework.integration:type=MessageHandler,name=throughputSink.throughputSink.serviceActivator,bean=endpoint': registering with JMX server as MBean [org.springframework.integration:type=MessageHandler,name=throughputSink.throughputSink.serviceActivator,bean=endpoint]
2018-10-09 14:11:27.927  INFO 3107 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase -2147483648
2018-10-09 14:11:27.928  INFO 3107 --- [           main] o.s.i.endpoint.EventDrivenConsumer       : Adding {service-activator:throughputSink.throughputSink.serviceActivator} as a subscriber to the 'input' channel
2018-10-09 14:11:27.928  INFO 3107 --- [           main] o.s.integration.channel.DirectChannel    : Channel 'throughtput-sink.input' has 1 subscriber(s).
2018-10-09 14:11:27.928  INFO 3107 --- [           main] o.s.i.endpoint.EventDrivenConsumer       : started throughputSink.throughputSink.serviceActivator
2018-10-09 14:11:27.928  INFO 3107 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase -2147482648
2018-10-09 14:11:27.934  INFO 3107 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase 0
2018-10-09 14:11:27.934  INFO 3107 --- [           main] o.s.i.endpoint.EventDrivenConsumer       : Adding {logging-channel-adapter:_org.springframework.integration.errorLogger} as a subscriber to the 'errorChannel' channel
2018-10-09 14:11:27.935  INFO 3107 --- [           main] o.s.i.channel.PublishSubscribeChannel    : Channel 'throughtput-sink.errorChannel' has 1 subscriber(s).
2018-10-09 14:11:27.935  INFO 3107 --- [           main] o.s.i.endpoint.EventDrivenConsumer       : started _org.springframework.integration.errorLogger
2018-10-09 14:11:27.935  INFO 3107 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase 2147482647
2018-10-09 14:11:28.041  INFO 3107 --- [           main] s.c.a.AnnotationConfigApplicationContext : Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@6ed3ccb2: startup date [Tue Oct 09 14:11:28 UTC 2018]; parent: org.springframework.context.annotation.AnnotationConfigApplicationContext@5c0369c4
2018-10-09 14:11:29.441  INFO 3107 --- [           main] o.s.c.s.b.k.p.KinesisStreamProvisioner   : Using Kinesis stream for inbound: devTimeSource
2018-10-09 14:11:29.563  INFO 3107 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'throughtput-sink-1.devTimeSource.throughtput-sink.errors' has 1 subscriber(s).
2018-10-09 14:11:29.571  INFO 3107 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'throughtput-sink-1.devTimeSource.throughtput-sink.errors' has 0 subscriber(s).
2018-10-09 14:11:29.572  INFO 3107 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'throughtput-sink-1.devTimeSource.throughtput-sink.errors' has 1 subscriber(s).
2018-10-09 14:11:29.572  INFO 3107 --- [           main] o.s.c.stream.binder.BinderErrorChannel   : Channel 'throughtput-sink-1.devTimeSource.throughtput-sink.errors' has 2 subscriber(s).
2018-10-09 14:11:29.599  INFO 3107 --- [           main] a.i.k.KinesisMessageDrivenChannelAdapter : started KinesisMessageDrivenChannelAdapter{shardOffsets=[KinesisShardOffset{iteratorType=TRIM_HORIZON, sequenceNumber='null', timestamp=null, stream='devTimeSource', shard='shardId-000000000000', reset=false}], consumerGroup='throughtput-sink'}
2018-10-09 14:11:29.608  INFO 3107 --- [           main] o.s.c.s.m.t.ThroughputSinkApplication    : Started ThroughputSinkApplication in 9.455 seconds (JVM running for 10.238)
2018-10-09 14:11:29.851  INFO 3107 --- [esis-consumer-1] a.i.k.KinesisMessageDrivenChannelAdapter : The [ShardConsumer{shardOffset=KinesisShardOffset{iteratorType=TRIM_HORIZON, sequenceNumber='null', timestamp=null, stream='devTimeSource', shard='shardId-000000000000', reset=false}, state=NEW}] has been started.
2018-10-09 14:11:35.326  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:       1712 in  5.00s =      342.40/s
2018-10-09 14:11:40.329  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        386 in  5.00s =       77.20/s
2018-10-09 14:11:45.330  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        350 in  5.00s =       70.00/s
2018-10-09 14:11:50.331  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        345 in  5.00s =       69.00/s
2018-10-09 14:11:55.331  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        306 in  5.00s =       61.20/s
2018-10-09 14:12:00.332  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        422 in  5.00s =       84.40/s
2018-10-09 14:12:05.333  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        352 in  5.00s =       70.40/s
2018-10-09 14:12:10.334  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        420 in  5.00s =       84.00/s
2018-10-09 14:12:15.334  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        315 in  5.00s =       63.00/s
2018-10-09 14:12:20.335  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        389 in  5.00s =       77.80/s
2018-10-09 14:12:25.336  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        283 in  5.00s =       56.60/s
2018-10-09 14:12:30.337  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        360 in  5.00s =       72.00/s
2018-10-09 14:12:35.337  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        409 in  5.00s =       81.80/s
2018-10-09 14:12:40.338  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        343 in  5.00s =       68.60/s
2018-10-09 14:12:45.339  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        377 in  5.00s =       75.40/s
2018-10-09 14:12:50.340  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        276 in  5.00s =       55.20/s
2018-10-09 14:12:55.340  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        334 in  5.00s =       66.80/s
2018-10-09 14:13:00.341  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        409 in  5.00s =       81.80/s
2018-10-09 14:13:05.342  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        411 in  5.00s =       82.20/s
2018-10-09 14:13:10.343  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        303 in  5.00s =       60.60/s
2018-10-09 14:13:15.343  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        388 in  5.00s =       77.60/s
2018-10-09 14:13:20.344  INFO 3107 --- [pool-1-thread-1] o.s.c.s.m.throughput.ThroughputSink      : Messages:        228 in  5.00s =       45.60/s
```
As you can see, we got less than 100 messages per second so 10 times less that we expect ! I would have expected a continuous flow of 1000 messages per second continuously. In local testing, we observe that from time to time, we do not receive any messages through the emulated Kinesis stream. We believe the **problem is the source configuration**. We have very low values thant we expected when we check the Kinesis metrics :
![image](https://user-images.githubusercontent.com/32193139/46723726-c6e77700-cc78-11e8-8a90-2dcb2811db44.png)

These are the configuration used by our modules :

time-source-kinesis producer:
```YAML
cloud:
  aws:
    region:
      static: eu-west-1

spring:
  application:
    name: time-source-kinesis
  cloud:
    stream:
      bindings:
        output:
          destination: devTimeSource
          binder: kinesis
          content-type: application/json
      kinesis:
        binder:
          checkpoint:
            table: devKinesisCheckPoints
            read-capacity: 10
            write-capacity: 10
          locks:
            table: devKinesisLocks
            read-capacity: 10
            write-capacity: 10

trigger:
  fixed-delay: 1
  time-unit: seconds
  max-messages: 1000
```

throughput-sink-kinesis consumer:

```YAML
cloud:
  aws:
    region:
      static: eu-west-1

spring:
  application:
    name: throughtput-sink
  cloud:
    stream:
      bindings:
        input:
          binder: kinesis
          destination: devTimeSource
          group: ${spring.application.name}
      kinesis:
        bindings:
          input:
            consumer:
              idleBetweenPolls: 250
        binder:
          checkpoint:
            table: devKinesisCheckPoints
            readCapacity: 10
            writeCapacity: 10
          locks:
            table: devKinesisLocks
            readCapacity: 10
            writeCapacity: 10
              
reportEveryMs: 5000
```

I may have forget a configuration properties but I do not know which. Thanks !

**Comment by artembilan** [comment]
Can you really confirm that this is an issue with the Kinesis Binder, but not an AWS protocol limitation?
For example in addition to that `idleBetweenPolls` we also have a:
```
this.consumerBackoff = Math.max(1000, consumerBackoff);
```
Which is used in this cases:
```
	if (ConsumerState.STOP != this.state && result.getRecords().isEmpty()) {
							if (logger.isDebugEnabled()) {
								logger.debug("No records for [" + this +
										"] on sequenceNumber [" +
										this.checkpointer.getLastCheckpointValue() +
										"]. Suspend consuming for [" +
										KinesisMessageDrivenChannelAdapter.this.consumerBackoff + "] milliseconds.");
							}
							prepareSleepState();
						}
```

If we don't have records in the shard interator, we have to sleep for at least one second. This is AWS recommendation though:
https://docs.aws.amazon.com/streams/latest/dev/troubleshooting-consumers.html

Plus need to take into account that records are not appeared for reading in the shard immediately, but I can't find the genuine doc on AWS now...

So, please, be more specific about where we really have problem.
We are fully opened for improvements, if you can point for us what, where and how we can fix. 

Thanks

**Comment by artembilan** [comment]
Also, take a look here, please: https://github.com/spring-cloud/spring-cloud-stream-binder-aws-kinesis/issues/44

**Comment by Brico87** [comment]
Hi, 
I looked at the stackoverflow topic on slow consumer. I already had set the `idleBetweenPolls` at 250ms on my consumer side. I believe my issue is linked with the **messages production**.

Let me give you some details about the message producer code. The maven parent is a `spring-boot-starter-parent` at version **2.0.5.RELEASE**. The dependencies versions are :
- spring-cloud : Finchley.SR1,
- spring-cloud-stream: Elmhurst.SR1,
- spring-cloud-aws: 2.0.0.RELEASE,
- spring-cloud-stream-binder-kinesis:1.0.0.RELEASE.

I am using the following spring boot app and message source connector : 
**MessageSourceApp.java**
```java
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.EnableConfigurationProperties;

@SpringBootApplication
@EnableConfigurationProperties
public class MessageSourceApp {
	public static void main(String[] args) {
		SpringApplication.run(MessageSourceApp.class, args);
	}
}
```
**MessageSourceConnector.java**
```java
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.cloud.stream.annotation.EnableBinding;
import org.springframework.cloud.stream.messaging.Source;
import org.springframework.context.annotation.Bean;
import org.springframework.integration.annotation.InboundChannelAdapter;
import org.springframework.integration.annotation.Poller;
import org.springframework.integration.core.MessageSource;
import org.springframework.messaging.support.GenericMessage;

import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.MeterRegistry;

@EnableBinding(Source.class)
public class MessageSourceConnector {

	/** The Constant LOGGER. */
	private static final Logger LOGGER = LoggerFactory.getLogger(MessageSourceConnector.class);

	private DateFormat dateFormat;
	private Counter nbMessagesEmitted;

	public MessageSourceConnector(MeterRegistry registry) {
		this.dateFormat = new SimpleDateFormat("yyyy-mm-dd hh:mm:ss");
		this.nbMessagesEmitted = Counter.builder("poc.message.source.nb.messages.emitted").register(registry);
	}

	@Bean
	@InboundChannelAdapter(value = Source.OUTPUT, poller = @Poller(fixedDelay = "${source.fixed-delay-ms}", maxMessagesPerPoll = "${source.number-messages}"))
	public MessageSource<String> timerMessageSource() {
		//@formatter:off
		return () -> {
			nbMessagesEmitted.increment();
			return new GenericMessage<>(getMessage());
		};
		//@formatter:on
	}

	private String getMessage() {
		return dateFormat.format(new Date());
	}
}
```

The `application.yml` file is the following one :
```yaml
server:
  port: 9061
management:
  server:
    port: 9062
  endpoints:
    web:
      exposure:
        include: "*"
cloud:
  aws:
    region:
      static: eu-west-1

source:
  fixed-delay-ms: 1000
  number-messages: 100

spring:
  application:
    name: message-source-kinesis
  cloud:
    stream:
      bindings:
        output:
          destination: devKinesisStream
          binder: kinesis
          content-type: application/json
      kinesis:
        binder:
          checkpoint:
            table: devKinesisCheckPoints
            read-capacity: 10
            write-capacity: 10
          locks:
            table: devKinesisLocks
            read-capacity: 10
            write-capacity: 10
```

When I am launching in debug mode, I have a series of logs like this one :
```bash
2018-10-11 08:16:28.691 DEBUG 17564 --- [ask-scheduler-2] o.s.i.e.SourcePollingChannelAdapter      : Poll resulted in Message: GenericMessage [payload=2018-16-11 08:16:28, headers={id=add48b8a-493a-fa01-2c5b-a20c78e94750, timestamp=1539245788691}]
2018-10-11 08:16:28.691 DEBUG 17564 --- [ask-scheduler-2] o.s.integration.channel.DirectChannel    : preSend on channel 'output', message: GenericMessage [payload=2018-16-11 08:16:28, headers={id=add48b8a-493a-fa01-2c5b-a20c78e94750, timestamp=1539245788691}]
2018-10-11 08:16:28.691 DEBUG 17564 --- [ask-scheduler-2] o.s.i.a.outbound.KinesisMessageHandler   : org.springframework.integration.aws.outbound.KinesisMessageHandler@237832e4 received message: GenericMessage [payload=byte[55], headers={contentType=application/json, id=fa503839-f86b-915c-000c-fd2e8aefe62f, timestamp=1539245788691}]
2018-10-11 08:16:28.691 DEBUG 17564 --- [ask-scheduler-2] o.s.integration.channel.DirectChannel    : postSend (sent=true) on channel 'output', message: GenericMessage [payload=byte[19], headers={contentType=application/json, id=deaec2b8-00e1-8aad-c595-6e5cca34cc02, timestamp=1539245788691}]
```

For each message generated, I see that we are calling the _handleMessageToAws_ function of the `spring-integration-aws` **KinesisMessageHandler.java**. We observe an average latency of the PutRecord operation of 6ms :
![image](https://user-images.githubusercontent.com/32193139/46792342-71c46780-cd43-11e8-98b1-60022ab1de60.png)

This is the first limitation observed: the use of the PutRecord function for each message generated in `spring-integration-aws`.

When I am launching in debug mode the throughput sink, I do not see the "No records for ..." message. Instead, I have a series of logs like this one: 
```bash
2018-10-11 07:45:30.059 DEBUG 18110 --- [esis-consumer-1] o.s.integration.channel.DirectChannel    : preSend on channel 'input', message: GenericMessage [payload=byte[55], headers={id=aa78baae-028a-b043-ba12-db4420cdb556, aws_shard=shar dId-000000000000, aws_receivedPartitionKey=1076125057, aws_receivedStream=devKinesisStream, aws_receivedSequenceNumber=49589077749387029240071309401219776072682716906194468866, timestamp=1539243930059}]
2018-10-11 07:45:30.059 DEBUG 18110 --- [esis-consumer-1] o.s.i.handler.ServiceActivatingHandler   : ServiceActivator for [org.springframework.integration.handler.MethodInvokingMessageProcessor@4f69264] (throughputSink.throughputSink.se rviceActivator.handler) received message: GenericMessage [payload=byte[19], headers={aws_shard=shardId-000000000000, id=d6b2e9ac-dc9d-c43e-537a-f9a4df156923, contentType=application/json, aws_receivedPartitionKey=1076125057, aws_receive dStream=devKinesisStream, aws_receivedSequenceNumber=49589077749387029240071309401219776072682716906194468866, timestamp=1539243930059}]
2018-10-11 07:45:30.059 DEBUG 18110 --- [esis-consumer-1] o.s.i.handler.ServiceActivatingHandler   : handler 'ServiceActivator for [org.springframework.integration.handler.MethodInvokingMessageProcessor@4f69264] (throughputSink.throughp utSink.serviceActivator.handler)' produced no reply for request Message: GenericMessage [payload=byte[19], headers={aws_shard=shardId-000000000000, id=d6b2e9ac-dc9d-c43e-537a-f9a4df156923, contentType=application/json, aws_receivedParti tionKey=1076125057, aws_receivedStream=devKinesisStream, aws_receivedSequenceNumber=49589077749387029240071309401219776072682716906194468866, timestamp=1539243930059}]
2018-10-11 07:45:30.059 DEBUG 18110 --- [esis-consumer-1] o.s.integration.channel.DirectChannel    : postSend (sent=true) on channel 'input', message: GenericMessage [payload=byte[19], headers={aws_shard=shardId-000000000000, id=d6b2e9a c-dc9d-c43e-537a-f9a4df156923, contentType=application/json, aws_receivedPartitionKey=1076125057, aws_receivedStream=devKinesisStream, aws_receivedSequenceNumber=49589077749387029240071309401219776072682716906194468866, timestamp=153924 3930059}]
```

So I am receiving correctly all messages pushed into the stream.

What is telling me that the problem is not from the consumer side, it is this log :
![image](https://user-images.githubusercontent.com/32193139/46792702-3fffd080-cd44-11e8-8a87-a8a0f05e9b73.png)

When I initialy push some messages into the stream, then stop the producer, start the consumer and then start the producer again, we can see that we have a huge throughput linked to the messages already present in the stream and then we have a **slow consumption linked with** what I believe is **a slow production**.

The `application.yml` file of the throughtput sink app is this one :
```yaml
cloud:
  aws:
    region:
      static: eu-west-1

spring:
  application:
    name: throughtput-sink
  cloud:
    stream:
      bindings:
        input:
          binder: kinesis
          destination: devKinesisStream
          group: ${spring.application.name}
      kinesis:
        bindings:
          input:
            consumer:
              idleBetweenPolls: 250
              consumer-backoff: 1000
        binder:
          checkpoint:
            table: devKinesisCheckPoints
            readCapacity: 10
            writeCapacity: 10
          locks:
            table: devKinesisLocks
            readCapacity: 10
            writeCapacity: 10
              
reportEveryMs: 5000
```

I will do some changes in my code in order to do some tracing from producer to consumer and I will try to implement the KPL/KCL to see if the issue is really linked with the Kinesis binder. 

I still think that I am using the binder not in the correct way but I am not able to see what I am missing ! Thanks.

**Comment by artembilan** [comment]
For the `PutRecords` request see, please, this SO question and its answer: https://stackoverflow.com/questions/52555914/spring-cloud-stream-producer-batch

**Comment by artembilan** [comment]
>  Kinesis appears to provide us messages after a very long time.

That was my point when I mentioned all those AWS Kinesis limitations.
I don't think that we are going to mock it somehow on the consumer side.

What we may do as something like really `batchMode` on the produce side. Only need to figure out how to implement it...
Maybe something like an `aggregator` between `Source.OUTPUT` and the target `KinesisMessageHandler` to really build that `PutRecordsRequest` automatically.
Only what is bad from here that during aggregation we may stop application or it can simply fail, so we are going to lose producing data. 
I don't think that storing intermediate messages for an aggregation somewhere in the DynamoDB is going to be a performance benefit for us.
That's why it might be better to deffer such a feature to the end-user it his/her Source code...

WDYT?

**Comment by Brico87** [comment]
We can indeed create an aggregator on the application side but we just miss the creation mecanism of a single `PutRecordsRequest` in order to aggregate them. 

In the `KinesisMessageHandler`, there is the `buildPutRecordRequest` function which is building a PutRecordRequest object from a **Message** object. I am going to implement my own using your answer in the _spring-cloud-stream-producer-batch_ question in SO in order to build a `PutRecordsRequest` from a list of data. Sadly, we are losing the JSON encoding/decoding mecanism in the `batch` process.

**Comment by artembilan** [comment]
What I mean with that my comment that you just need to do something like this:
```
@Bean
@InboundChannelAdapter(value = Source.OUTPUT, poller = @Poller(fixedDelay = "${source.fixed-delay-ms}", maxMessagesPerPoll = "${source.number-messages}"))
public MessageSource<PutRecordsRequest> timerMessageSource() {
   ...
}
```

Pay attention to the return type. 
Well, you can perform JSON encoding yourself, using know Jackson `ObjectMapper`.
We don't have choice with the `PutRecordsRequest`: it expects from us a set of `PutRecordsRequestEntry` based on the `ByteBuffer`.

**Comment by r1m** [comment]
Hi artem,
Regarding your comment on aggregation failed issue, how does the kafka or rabbit binder handles the batch publishing ?

**Comment by artembilan** [comment]
Seems for me they just don't have such a feature and just perform regular publish-async-per-record.

I will let @garyrussell to comment on the matter. I'm not sure that I'm fully on board with internal features of Kafka and Rabbit Binders...

**Comment by garyrussell** [comment]
@r1m @artembilan 

For the kafka binder, the kafka client can be configured to batch messages, based on the `linger.ms` and `batch.size` kafka producer properties - it works like Nagle's algorithm in TCP; the client will wait for that time to elapse before sending. See the [kafka docs](https://kafka.apache.org/documentation/).

For the Rabbit Binder, a similar mechanism is provided by [Spring AMQP](https://docs.spring.io/spring-amqp//reference/html/_reference.html#template-batching), enabled by the `batchingEnabled`, `batchSize`, `batchBufferLimit` and `batchTimeout` producer properties.

**Comment by artembilan** [comment]
Thanks, @garyrussell , for sharing such a valuable info!
Doesn't those batching features there mean that we may lose data if app fail in between?
That was always a bottleneck when we deal with an `aggregator`...

**Comment by garyrussell** [comment]
That is correct - trade off between performance and reliability.

**Comment by artembilan** [comment]
That's fair.

So, for this issue we can simply go ahead and insert an `AggregatingMessageHandler` in between an output binding and the mentioned `KinesisMessageHandler` on the producer side when new `producerMode` property is set to `true`.

Does it sound like a plan, @r1m , @garyrussell  ?

**Comment by garyrussell** [comment]
tl;dr; (the whole thread).

I am not sure how an aggregator will help; if it's configured to emit a group in a single message, the consumer will receive a group, not discrete messages. If it's configured to emit discrete messages, won't they be sent one at a time as before?

**Comment by artembilan** [comment]
@garyrussell ,

The point is that Kinesis has a batching feature with its `PutRecordsRequest`. So, an idea is to buffer source messages in the memory and build such a container object by some provided size with the same correlation id.
For plain (single record mode) we are going to send a `PutRecordRequest`.

**Comment by r1m** [comment]
Hello Artem,
With @Brico87, we managed to create an aggregator that send `PutRecordsRequest`.
Main issues are about AWS limitations; it's hard to be below the 1000 records/shard/s **AND** the 500 records/PutRecordsRequest limits : we don't know which shard will be used after partitioning.

This leads us to over-provision our shards. 
With fine tuning of the group size and release timing, we reduced the number but it's still not perfect.
Our observations are that we need twice the shard needed if we were using them 100% perfectly.

**Comment by artembilan** [comment]
@r1m ,

That's great that you have a progress with the `PutRecordsRequest` aggregation.
The limitation is known and we have nothing here to do with that: our goal is to aggregate requests into the group with maximum 500 size and allow to emit not full group using some (optional) group timeout.

Everything else with over-sharding is already out of scope of this project, although we have some hook on the matter with the `int shardCount = properties.getInstanceCount() * properties.getConcurrency();` logic, where those options are part of the `ConsumerProperties` configuration: https://docs.spring.io/spring-cloud-stream/docs/current/reference/htmlsingle/#_consumer_properties
